File: python/ccxt/moomoo.py
# -*- coding: utf-8 -*-

# -----------------------------------------------------------------------------

from ccxt.base.exchange import Exchange
from ccxt.base.errors import ArgumentsRequired, NotSupported

# -----------------------------------------------------------------------------


class moomoo(Exchange):
    """Broker adapter scaffold for Moomoo. Configure options['brokerAdapter'] with deterministic callbacks to bridge the vendor SDK/API."""

    def describe(self):
        return self.deep_extend(super(moomoo, self).describe(), {
            'id': 'moomoo',
            'name': 'Moomoo',
            'countries': ['US', 'HK', 'SG'],
            'rateLimit': 1000,
            'certified': False,
            'pro': False,
            'has': {
                'CORS': None,
                'spot': True,
                'margin': True,
                'swap': False,
                'future': False,
                'option': True,
                'createOrder': True,
                'cancelOrder': True,
                'fetchBalance': True,
                'fetchMarkets': 'emulated',
                'fetchMyTrades': True,
                'fetchOpenOrders': True,
                'fetchOrder': True,
                'fetchOrderBook': True,
                'fetchPositions': True,
                'fetchTicker': True,
                'fetchTrades': True,
            },
            'urls': {
                'www': 'https://www.moomoo.com',
                'doc': ['https://openapi.moomoo.com/'],
                'api': {
                    'broker': 'moomoo://local-adapter',
                },
            },
            'requiredCredentials': {
                'apiKey': False,
                'secret': False,
                'uid': False,
                'accountId': False,
            },
            'options': {
                'markets': [],
                'brokerAdapter': None,
            },
        })

    def call_broker(self, method, params={}):
        adapter = self.safe_value(self.options, 'brokerAdapter')
        if adapter is None:
            raise NotSupported(self.id + ' requires options.brokerAdapter.' + method + ' for this operation')
        callback = adapter.get(method) if isinstance(adapter, dict) else getattr(adapter, method, None)
        if callback is None:
            raise NotSupported(self.id + ' requires options.brokerAdapter.' + method + ' for this operation')
        return callback(params)

    def fetch_markets(self, params={}):
        return self.safe_value(self.options, 'markets', [])

    def fetch_balance(self, params={}):
        response = self.call_broker('fetchBalance', params)
        return self.safe_balance(response)

    def fetch_ticker(self, symbol, params={}):
        if symbol is None:
            raise ArgumentsRequired(self.id + ' fetch_ticker() requires a symbol argument')
        self.load_markets()
        market = self.market(symbol)
        request = {'symbol': symbol, 'market': market}
        response = self.call_broker('fetchTicker', self.extend(request, params))
        return self.safe_ticker(response, market)

    def fetch_order_book(self, symbol, limit=None, params={}):
        self.load_markets()
        market = self.market(symbol)
        request = {'symbol': symbol, 'market': market, 'limit': limit}
        response = self.call_broker('fetchOrderBook', self.extend(request, params))
        return self.parse_order_book(response, market['symbol'])

    def create_order(self, symbol, type, side, amount, price=None, params={}):
        self.load_markets()
        market = self.market(symbol)
        request = {'symbol': symbol, 'market': market, 'type': type, 'side': side, 'amount': amount, 'price': price}
        response = self.call_broker('createOrder', self.extend(request, params))
        return self.parse_order(response, market)

    def cancel_order(self, id, symbol=None, params={}):
        response = self.call_broker('cancelOrder', self.extend({'id': id, 'symbol': symbol}, params))
        return self.parse_order(response)

    def fetch_order(self, id, symbol=None, params={}):
        response = self.call_broker('fetchOrder', self.extend({'id': id, 'symbol': symbol}, params))
        return self.parse_order(response)

    def fetch_open_orders(self, symbol=None, since=None, limit=None, params={}):
        request = {'symbol': symbol, 'since': since, 'limit': limit}
        response = self.call_broker('fetchOpenOrders', self.extend(request, params))
        return self.parse_orders(response, None, since, limit)

    def fetch_my_trades(self, symbol=None, since=None, limit=None, params={}):
        request = {'symbol': symbol, 'since': since, 'limit': limit}
        response = self.call_broker('fetchMyTrades', self.extend(request, params))
        return self.parse_trades(response, None, since, limit)

    def fetch_trades(self, symbol, since=None, limit=None, params={}):
        request = {'symbol': symbol, 'since': since, 'limit': limit}
        response = self.call_broker('fetchTrades', self.extend(request, params))
        return self.parse_trades(response, None, since, limit)

    def fetch_positions(self, symbols=None, params={}):
        response = self.call_broker('fetchPositions', self.extend({'symbols': symbols}, params))
        return self.parse_positions(response, symbols)

File: ts/src/moomoo.ts
// ---------------------------------------------------------------------------

import Exchange from './base/Exchange.js';
import { ArgumentsRequired, NotSupported } from './base/errors.js';
import type { Balances, Dict, Market, Num, Order, OrderBook, OrderSide, OrderType, Str, Ticker, Trade, int } from './base/types.js';

// ---------------------------------------------------------------------------

/**
 * @class moomoo
 * @augments Exchange
 * @description Broker adapter scaffold for Moomoo. Configure options.brokerAdapter with deterministic callbacks to bridge the vendor SDK/API.
 */
export default class moomoo extends Exchange {
    describe (): any {
        return this.deepExtend (super.describe (), {
            'id': 'moomoo',
            'name': 'Moomoo',
            'countries': [ 'US', 'HK', 'SG' ],
            'rateLimit': 1000,
            'certified': false,
            'pro': false,
            'has': {
                'CORS': undefined,
                'spot': true,
                'margin': true,
                'swap': false,
                'future': false,
                'option': true,
                'createOrder': true,
                'cancelOrder': true,
                'fetchBalance': true,
                'fetchMarkets': 'emulated',
                'fetchMyTrades': true,
                'fetchOpenOrders': true,
                'fetchOrder': true,
                'fetchOrderBook': true,
                'fetchPositions': true,
                'fetchTicker': true,
                'fetchTrades': true,
            },
            'urls': {
                'www': 'https://www.moomoo.com',
                'doc': [ 'https://openapi.moomoo.com/' ],
                'api': {
                    'broker': 'moomoo://local-adapter',
                },
            },
            'requiredCredentials': {
                'apiKey': false,
                'secret': false,
                'uid': false,
                'accountId': false,
            },
            'options': {
                // Markets can be provided by the embedding app when the vendor API does not expose a CCXT-style catalogue.
                'markets': [],
                // brokerAdapter methods: fetchBalance, fetchTicker, fetchOrderBook, createOrder, cancelOrder,
                // fetchOpenOrders, fetchOrder, fetchMyTrades, fetchTrades, fetchPositions.
                'brokerAdapter': undefined,
            },
        });
    }

    async callBroker (method: string, params: Dict = {}): Promise<any> {
        const adapter = this.safeValue (this.options, 'brokerAdapter');
        if ((adapter === undefined) || (adapter[method] === undefined)) {
            throw new NotSupported (this.id + ' requires options.brokerAdapter.' + method + ' for this operation');
        }
        return await adapter[method] (params);
    }

    async fetchMarkets (params = {}): Promise<Market[]> {
        const markets = this.safeValue (this.options, 'markets', []);
        return markets as Market[];
    }

    async fetchBalance (params = {}): Promise<Balances> {
        const response = await this.callBroker ('fetchBalance', params);
        return this.safeBalance (response);
    }

    async fetchTicker (symbol: string, params = {}): Promise<Ticker> {
        if (symbol === undefined) {
            throw new ArgumentsRequired (this.id + ' fetchTicker() requires a symbol argument');
        }
        await this.loadMarkets ();
        const market = this.market (symbol);
        const response = await this.callBroker ('fetchTicker', this.extend ({ 'symbol': symbol, 'market': market }, params));
        return this.safeTicker (response, market);
    }

    async fetchOrderBook (symbol: string, limit: int = undefined, params = {}): Promise<OrderBook> {
        await this.loadMarkets ();
        const market = this.market (symbol);
        const request = { 'symbol': symbol, 'market': market, 'limit': limit };
        const response = await this.callBroker ('fetchOrderBook', this.extend (request, params));
        return this.parseOrderBook (response, market['symbol']);
    }

    async createOrder (symbol: string, type: OrderType, side: OrderSide, amount: Num, price: Num = undefined, params = {}): Promise<Order> {
        await this.loadMarkets ();
        const market = this.market (symbol);
        const request = {
            'symbol': symbol,
            'market': market,
            'type': type,
            'side': side,
            'amount': amount,
            'price': price,
        };
        const response = await this.callBroker ('createOrder', this.extend (request, params));
        return this.parseOrder (response, market);
    }

    async cancelOrder (id: string, symbol: Str = undefined, params = {}): Promise<Order> {
        const request = { 'id': id, 'symbol': symbol };
        const response = await this.callBroker ('cancelOrder', this.extend (request, params));
        return this.parseOrder (response);
    }

    async fetchOrder (id: string, symbol: Str = undefined, params = {}): Promise<Order> {
        const request = { 'id': id, 'symbol': symbol };
        const response = await this.callBroker ('fetchOrder', this.extend (request, params));
        return this.parseOrder (response);
    }

    async fetchOpenOrders (symbol: Str = undefined, since: int = undefined, limit: int = undefined, params = {}): Promise<Order[]> {
        const request = { 'symbol': symbol, 'since': since, 'limit': limit };
        const response = await this.callBroker ('fetchOpenOrders', this.extend (request, params));
        return this.parseOrders (response, undefined, since, limit);
    }

    async fetchMyTrades (symbol: Str = undefined, since: int = undefined, limit: int = undefined, params = {}): Promise<Trade[]> {
        const request = { 'symbol': symbol, 'since': since, 'limit': limit };
        const response = await this.callBroker ('fetchMyTrades', this.extend (request, params));
        return this.parseTrades (response, undefined, since, limit);
    }

    async fetchTrades (symbol: string, since: int = undefined, limit: int = undefined, params = {}): Promise<Trade[]> {
        const request = { 'symbol': symbol, 'since': since, 'limit': limit };
        const response = await this.callBroker ('fetchTrades', this.extend (request, params));
        return this.parseTrades (response, undefined, since, limit);
    }

    async fetchPositions (symbols: Str[] = undefined, params = {}): Promise<any[]> {
        const response = await this.callBroker ('fetchPositions', this.extend ({ 'symbols': symbols }, params));
        return this.parsePositions (response, symbols);
    }
}
